apps/namadillo/src/App/Swap/state/atoms.ts

```diff
@@ -0,0 +1,182 @@
+      sellAsset: baseState.sellAsset || sellAsset,
+      buyAsset: baseState.buyAsset || buyAsset,
+    };
+  },
+  (get, set, update: SwapState | ((prev: SwapState) => SwapState)) => {
+    set(baseSwapStateAtom, update);
+    const { sellAsset, buyAsset } = get(internalSwapStateAtom);
+
+    set(sellAssetAtom, sellAsset?.symbol);
+    set(buyAssetAtom, buyAsset?.symbol);
+  }
+);
+
+// We use debounced version of the swap state for fetching quotes to avoid excessive requests
+const debouncedSwapStateAtom = atom<SwapState>(defaultSwapState);
+
+const debouncedSet = debounce((set: Setter, value: SwapState) => {
+  set(debouncedSwapStateAtom, value);
+}, 300);
+
+export const swapStateAtom = atom(
+  (get) => get(internalSwapStateAtom),
+  (get, set, update: SwapState | ((prev: SwapState) => SwapState)) => {
+    set(internalSwapStateAtom, update);
+    const val = get(internalSwapStateAtom);
+    debouncedSet(set, val);
+  }
+);
+
+export const setInternalSwapStateAtom = atom(
+  null,
+  (_get, set, update: SwapState | ((prev: SwapState) => SwapState)) => {
+    set(internalSwapStateAtom, update);
+  }
+);
+
+export const swapStatusAtom = atom<SwapStatusType>(SwapStatus.idle());
+
+export const swapQuoteAtom = atomWithQuery((get) => {
+  const swapState = get(debouncedSwapStateAtom);
+  const { sellAsset, buyAsset } = swapState;
+
+  // TODO: this should be configurable once we support osmosis testnet
+  const osmosisAssets =
+    getChainRegistryByChainId("osmosis-1")?.assets.assets || [];
+
+  //We only want to refetch when sellAmount changes when selling, and buyAmount when buying
+  const sellKey =
+    swapState.mode === "sell" ?
+      `${swapState.sellAmount}-${swapState.sellAsset?.symbol}`
+    : "sell";
+  const buyKey =
+    swapState.mode === "buy" ?
+      `${swapState.buyAmount}-${swapState.buyAsset?.symbol}`
+    : "buy";
+
+  return {
+    enabled: Boolean(sellAsset && buyAsset),
+    queryKey: ["swapQuote", sellKey, buyKey],
+    queryFn: async () => {
+      // Sanity checks
+      invariant(sellAsset, "Sell asset not found");
+      invariant(buyAsset, "Buy asset not found");
+
+      const fromOsmosis = osmosisAssets.find(
+        (assets) => assets.symbol === sellAsset.symbol
+      );
+      const toOsmosis = osmosisAssets.find(
+        (assets) => assets.symbol === buyAsset.symbol
+      );
+
+      invariant(fromOsmosis, "From asset is not found in Osmosis assets");
+      invariant(toOsmosis, "To asset is not found in Osmosis assets");
+
+      // If amount is empty, we still want to get a quote for 1 unit of the asset
+      const baseAmount =
+        swapState.mode === "sell" ?
+          toBaseAmount(sellAsset, swapState.sellAmount!)
+        : swapState.mode === "buy" ?
+          toBaseAmount(buyAsset, swapState.buyAmount!)
+        : toBaseAmount(buyAsset, BigNumber(1));
+
+      const simulateSell =
+        swapState.mode === "sell" || swapState.mode === "none";
+
+      const params: Record<string, string> =
```

Comment:
can remove this casting and update the type in `fetchQuote` if you want....up to you